[TensorExpr] Bounds inference

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36112 [TensorExpr] Implement LoopNest::computeAt
* **#35120 [TensorExpr] Bounds inference**



Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

Commit:
[NOT READY] Bounds inference

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
clang-format on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[NOT READY][TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
clang-tidy on "[TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

nickgg commented on torch/csrc/jit/tensorexpr/bounds_inference.cpp:
```diff
@@ -0,0 +1,172 @@
+
+static const Expr* simplifyMax(const Expr* a, const Expr* b) {
+  const Expr* diff = IRSimplifier::simplify(new Sub(a, b));
+  if (auto diff_imm = dynamic_cast<const IntImm*>(diff)) {
```
This should probably be `isConstant`, although it feels like using the inferred bounds could get us a little closer in some instances.

zheng-xq commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -546,5 +547,224 @@ void testScheduleDynamicShape2D() {
+  Tensor* b =
+      Compute("b", {{n, "i"}}, [&](const VarHandle& i) { return a(i); });
+  LoopNest l({b});
+  const std::vector<TensorAccessBoundsInfo>& bounds_info =
```
As for the interface of bounds inference module, those API should be abstracted to encapsulate its implementations.

TensorAccessBounds can be an object itself. Whether the data structure is a vector, or a map can be hidden away from its user.

"bounds_info_map.at(a.data())" Can be replaced by a convenience function "tensor_bounds.find(a)".

ZolotukhinM replied:
The API is to return a collection (in this implementation a vector) of `TensorAccessBoundsInfo` objects - I don't see a value in wrapping the collection into a separate class when it's nothing else than just a vector. Note that everyone who reads this code would know what `std::vector` is, but noone except me would know what `TensorAccessBounds` class is.

zheng-xq replied:
Hiding away the implementation details is a key idea of software abstraction. The interface of this module should reflect its logic usage. If we change the internal data structure, we don't have to change all the call sites, as this might be used in many places. Also we don't have to advertise to every developer the change of the implementation data structure. So interface between the modules can stay relatively stable.

zheng-xq commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -546,5 +547,224 @@ void testScheduleDynamicShape2D() {
+  // for i in 0..100:
+  //   b[i] = a[i]
+  // For this loop bounds inference should yield the following:
+  // {{b, kStore, 0, 99}, {a, kLoad, 0, 99}}
```
Minor: as for the bound, I mild prefer the C++ convention: left-close, right-open. So the same range should be: [0, 100)

ZolotukhinM replied:
It is actually quite inconvenient in this case to use that convention the reason for that is how bounds inference works: it starts from a scope of Load/Store/FunctionCall and records indices used there and then escapes to outer scopes one by one replacing start/stop values for the corresponding index with Min/Max values of the iteration range. E.g:
```
(1)  for x in 0..10:
(2)    a[x] = b[x*2]
```
In this loopnest will start from (2) and infer bounds: `b: [x*2; x*2], a: [x; x]`. Then we will go to the scope of the loop (1). We will take the inferred bounds and in them we will replace `x` with `0` for the start value and with `10-1` (the last taken value, which equals stop values minus 1) for the stop value: `b: [0;18], a:[0;9]`. We could add +1 everywhere to follow the `[start;stop)` notation, but we would still have to do this `-1` and then `+1` at every step: i.e. we cannot just replace `x` with `10` (stop value), as it would give us incorrect bounds `b:[0:20), a:[0:10)` (correct bounds for `b` would be `[0:19)`).

nickgg commented on torch/csrc/jit/tensorexpr/bounds_inference.cpp:
```diff
@@ -0,0 +1,172 @@
+static const Expr* simplifyMax(const Expr* a, const Expr* b) {
+  const Expr* diff = IRSimplifier::simplify(new Sub(a, b));
+  if (auto diff_imm = dynamic_cast<const IntImm*>(diff)) {
+    if (diff_imm->value() > 0) {
```
Then this could be `immediateIsNegative` - existing helper in ir.h which handles more dtypes

ZolotukhinM replied:
I'll move this part to IR simplifier in a follow-up diff if you don't mind.

zheng-xq commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -546,5 +547,224 @@ void testScheduleDynamicShape2D() {
+  // We should have two entries: one for 'b' and one for 'a'.
+  ASSERT_EQ(bounds_info_map.size(), 2);
+  ASSERT_EQ(bounds_info_map.at(a.data()).kind, kLoad);
+  verifyConstBounds(bounds_info_map.at(a.data()), {{0, -1}});
```
"-1" looks a bit weird to imply "None". Most system use "None", which is a particular very negative number. Or an OptionalNumber where it can be implicitly constructed from an integer, or None which implies that there is no number.

ZolotukhinM replied:
This is only for these three tests, I would prefer not to over-engineer this.

zheng-xq commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -546,5 +547,224 @@ void testScheduleDynamicShape2D() {
+  // for y in 0..200:
+  //   for x in 0..320:
+  //     b[y,x] = x*y
+  // for y in 0..200:
```
An important case is where the "y, x" in "c" uses a small range in "a" and "b".

ZolotukhinM replied:
Added!

nickgg reviewed (commented):
Yeah I think this makes sense as a part of the IR Simplifier, we could extend it to handle CompareSelect as well.

zheng-xq commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -546,5 +547,224 @@ void testScheduleDynamicShape2D() {
+  }
+}
+
+void testBoundsInference_5() {
```
Another important test case is where the same source tensor is used multiple times in the target tensor, potentially through different path: 

a = f(x, y)
b = a(x, y)
c = b(x+1, y)
d = a(x, y+1) + c(x, y)
compute(d)
a.compute_at(d, x)

ZolotukhinM replied:
That comment is probably to the next PR, not this one?

ZolotukhinM replied:
I added this test to `computeAt` tests.

Commit:
Update on "[TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Bounds inference"




Differential Revision: [D20567926](https://our.internmc.facebook.com/intern/diff/D20567926)

[ghstack-poisoned]
